format/spec.md

```diff
@@ -841,19 +855,45 @@ Notes:
+
+Both position and equality delete files allow encoding deleted row values with a delete. This can be used to reconstruct a stream of changes to a table.
+
 
-Row-level delete files are valid Iceberg data files: files must use valid Iceberg formats, schemas, and column projection. It is recommended that delete files are written using the table's default file format.
+### Deletion Vectors
 
-Row-level delete files are tracked by manifests, like data files. A separate set of manifests is used for delete files, but the manifest schemas are identical.
+Deletion vectors identify deleted rows of a file by encoding deleted positions in a bitmap. A set bit at position P indicates that the row at position P is deleted.
 
-Both position and equality deletes allow encoding deleted row values with a delete. This can be used to reconstruct a stream of changes to a table.
+These vectors are stored using the `delete-vector-v1` blob definition from the [Puffin spec][puffin-spec].
 
+Deletion vectors support positive 64-bit positions, but are optimized for cases where most positions fit in 32 bits by using a collection of 32-bit Roaring bitmaps. 64-bit positions are divided into a 32-bit "key" using the most significant 4 bytes and a 32-bit sub-position using the least significant 4 bytes. For each key in the set of positions, a 32-bit Roaring bitmap is maintained to store a set of 32-bit sub-positions for that key.
+
+To test whether a certain position is set, its most significant 4 bytes (the key) are used to find a 32-bit bitmap and the least significant 4 bytes (the sub-position) are tested for inclusion in the bitmap. If a bitmap is not found for the key, then it is not set.
+
+Delete manifests track deletion vectors individually by the containing file location (`file_path`), starting offset of the DV magic bytes (`blob_offset`), and total length of the deletion vector blob (`blob_size_in_bytes`). Multiple deletion vectors can be stored in the same file. There are no restrictions on the data files that can be referenced by deletion vectors in the same Puffin file.
+
+At most one deletion vector is allowed per data file in a table. If a DV is written for a data file, it must replace all previously written position delete files so that when a DV is present, readers can safely ignore matching position delete files.
+
+
+[puffin-spec]: https://iceberg.apache.org/puffin-spec/
 
 ### Position Delete Files
 
 Position-based delete files identify deleted rows by file and position in one or more data files, and may optionally contain the deleted row.
 
+_Note: Position delete files are **deprecated** in v3. Existing position deletes must be written to delete vectors when updating the position deletes for a data file._
```

Comment:
Is there a technical reason to force deprecation here?

Reply:
Yes! Deletion vectors are much better for creating a stream of changes from tables. They are also just faster in general so we want to switch over to using them.

Reply:
There are many optimizations enabled by knowing there is a DV for a data file, this was one of the points in the design doc.

Reply:
On the flip side a counter argument, is that requriing DVs means more work for writers that want to support some portions of V3 specification but not upgrade to DVs (despite them being better).  It seems other places iceberg has taken the stand that optimizations are not required for writers (e.g. sequence number inheritance).

Reply:
> optimizations are not required for writers (e.g. sequence number inheritance).

Could you elaborate? Not sure I got.

Reply:
My understanding is that Snapshot/Sequence number inheritance is only turned on via a configuration flag for Iceberg tables for writers.  It is not required that writers "must" make snapshot id null for new files added.  Similarly, I'm asking that if it really needs to be made a requirement that DVs must be written, even though they are better.

Reply:
Actually, Iceberg is fairly strict about writing correct and well-maintained metadata, but we are careful to be backward-compatible. Sequence number and snapshot ID inheritance are required in v2. I think what you're referring to here is that we made snapshot ID inheritance available in v1 with a flag to enable it (and break compatibility with some readers).

Reply:
I guess what I'm referring to in this case might have been considered a [bug that was addressed](https://github.com/apache/iceberg/commit/50c5f267b7f2919fec77f8f690412c0cb24b0de5).  But the spec has never mandated inheritance "must" be used.

I think the question still stands on whether from a spec perspective this should be mandated ("must") or ("should").  I guess I'm OK with "must"  given clarification below on existing delete files are present but it still feels like we might be overly strict.

Reply:
The snapshot and sequence number inheritance was always required in V2 tables from the spec perspective, it just took us a bit of time to get rid of the legacy V1 logic everywhere. The flag was there only to enable that behavior in V1 tables, which could break older readers. That's why it was safe to merge the referenced PR as it simply aligned the implementation with the spec.

We know the new representation is better. I think it is OK to be strict in this case. We were very flexible with the design of equality deletes. For instance, we allowed adding any number of global equality delete files. We thought  implementations would be smart and would not misuse this concept. Yet, we have seen so many cases where tables were unreadable after engines produces a ton of equality deletes without ever compacting them. I think the new layout forces implementations to provide the best possible behavior for position deletes as well as enables various optimizations like CDC generation mentioned by Ryan.